pkgs/top-level/all-packages.nix

```diff
@@ -4066,7 +4066,19 @@ in
 
   dleyna-server = callPackage ../development/libraries/dleyna-server { };
 
-  dmd = callPackage ../development/compilers/dmd { };
+  dmd_2_95 = callPackage ../development/compilers/dmd
+  { version = "2.095.1";
+    inherit (darwin.apple_sdk.frameworks) Foundation;
+  };
+  dmd_2_96 = callPackage ../development/compilers/dmd
+  { version = "2.096.1";
+    inherit (darwin.apple_sdk.frameworks) Foundation;
+  };
```

Comment:
Do we really need and plan to support all these versions?

Reply:
Hard to say. I could be convinced that it should be only every 2-5 versions in `allpackages.nix` instead, to avoid too much work for Hydra. But I definitely do think it should be more than just the version, because DMD versions are essentially "rolling releases" - new version every two or three months, with the language changing steadily.

D does have some backward compatibility, but when working with rarely-updated code it's still fairly common that you can't just use the newest version. Is there any standing policy about how frequent the versioning should be in cases like this?

Reply:
I think that every four versions, plus the latest stable, is a good frequency canditate. I changed the versions to 2.092.1, 2.096.1 and 2.097.2.

I would be happy to have 2.088 and 2.084 too but perhaps another day.

Reply:
Nixpkgs is not the best place to have multiple versions that you can reliably refer to in your projects. Unused packages can be removed at any time, or a specific version that you may need may not be available. If you really want a specific version, a local override could work better for you.

Reply:
I disagree. First off, we are talking about a compiler. It is common to have to compile some codebase with older compilers in just about any language. We do not want to require everyone with a bit outdated D codebase to make their own overloads. See https://forum.dlang.org/post/ecgwqlnepjqpjvixdypn@forum.dlang.org

Second, I think this scheme might even reduce the maintenance burden in the long run. There are already some D projects out there, such as the Tilix terminal emulator, and there may be more in the future. Without this scheme have to either:
-always update them when they stop compiling with newer compilers, or
-have them to use their own compiler overloads, which taken together are harder to maintain and slower to build than a common versioning scheme.

I'm open to discussing the detail though - how many older versions we should have and how big stride between them.

Reply:
I'm not saying that it can't be done. Only pointing at some long-term practical issues.

Reply:
Sorry but I also think that maintaining multiple versions is not a good idea.
Sometimes it really is the case that some project doesn't compile anymore with the newest compiler but that happens decreasingly often and it often is just a matter of pinging the project maintainers to update the code.
It's also possible to incorporate those changes into the nixpkgs until a real release is ready and if a project is so unmaintained that it's not getting fixed it can either be disabled, completely removed or fixed by an interested nixpkgs maintainer.
If somebody needs another version of the compiler for development it's very easy to build one itself via the already provided compiler from nixpkgs.
But for this scenario it would probably also make sense to just use something like DVM (https://code.dlang.org/packages/dvm) but I never tried if that works in Nix.

Regarding the maintenance burden. Out of experience I can tell that it was much less of a hassle to update all the depending packages than updating the dlang compilers. That has a lot to do with the fact that most of the tests are run for the compilers and that there are not that many packages in nixpkgs but I think that will not change so fast if ever.

I always tried to choose the latest stable version for dmd and LDC with the same frontend version when I updated the packages. Apart from this I would not care that much about the compiler version.

Reply:
I still think long-term-support versions would be worth it. But package maintainer is the best one to decide so I'm closing this unless Lionello wants to continue the discussion.

I'll resubmit 2.097.2 later as just a simple version bump.

Reply:
FWIW, I tend to agree with the others that nixpkgs should stick to a single version (at least, per major version). We should still try to make it as easy as possible for overlays/overrides to plug in custom versions (possibly browsing the GIT history to find the correct version-hash pairs.)

Reply:
I'm considering making my own repository where I will store the derivations for the older DMD builds. No precompiled binaries for the user, but otherwise almost as good as having them here.

Reply:
If anyone is still following, the said repository is done now: https://github.com/dukc/oldDDerivations